### ISSUE TITLE
indent error when work with new yasnippet-mode

### ISSUE THREAD

**Issue by zw963** [zenspider/enhanced-ruby-mode#158]
Following is a issue create from yasnippet repo, i just copy to here.

For reference, a quick reproducer, run as `emacs -Q -L .../yasnippet -L .../enh-ruby-mode -l issue-969.el`

```el
;;; issue-969.el

(require 'enh-ruby-mode)
(require 'yasnippet)
(require 'edebug)

(with-current-buffer (car (find-function-noselect 'yas--indent))
  (edebug-instrument-function 'yas--indent)
  (forward-sexp)
  (search-backward "yas--indent-region")
  (edebug-set-breakpoint nil))

(edebug-instrument-function 'enh-ruby-calculate-indent)
(setq edebug-initial-mode 'go)


(yas-define-snippets
 'enh-ruby-mode
 '(("do" "do${1:$(or \" |\" \"\")}${1:arg}${1:$(or \"|\" \"\")}
  $0
end")))

(with-current-buffer (get-buffer-create  "*yas test*")
  (erase-buffer)
  (enh-ruby-mode)
  (yas-minor-mode)
  (insert "method do")
  (pop-to-buffer (current-buffer)))
```

I don't really understand why it fails. For some reason
enh-ruby-calculate-indent returns 0 for the middle line when
initially expanding the snippet, but later it returns 2 as expected.
I find enh-ruby-mode's indentation code kind of hard to follow.

Following is original issue commit: 
https://github.com/joaotavora/yasnippet/issues/969#issuecomment-483067255

Thanks.

**Comment by npostavs** [comment]
I tracked down the yasnippet change to joaotavora/yasnippet@1d96da2e08664c31ff7f6f7441da1f4fa5680b1f, which is binding `inhibit-modification-hook` around snippet insertion. yasnippet does also do `(run-hook-with-args 'after-change-functions (point-min) (point-max) 0)` after the insertion is done (with narrowed around the snippet contents, that's why `(point-min)` and  `(point-max)` make sense), which I thought would be enough, but clearly that's not the case here. I tried adding

```elisp
(require 'trace)

(mapc #'trace-function '(process-send-string
                         erm-filter erm-req-parse))
```
to the reproducer above and comparing the traces in the good and bad cases, but I can't make much sense of the data sent to & from the ruby parser process.

**Comment by zenspider** [comment]
> I can't make much sense of the data sent to & from the ruby parser process.

Neither can I. 😕

I'm trying to slowly get some sort of doco going that'll define the wire protocol, but this stuff is pretty opaque.

**Comment by zenspider** [comment]
Can you pop in a call to `erm-wait-for-parse` before doing anything with `yasnippet` and see if that helps at all?

**Comment by npostavs** [comment]
> Can you pop in a call to `erm-wait-for-parse` before doing anything with `yasnippet` and see if that helps at all?

Yes, I can confirm the patch below fixes the issue:

```diff
--- i/yasnippet.el
+++ w/yasnippet.el
@@ -4164,6 +4164,9 @@ (defun yas--snippet-create (content expand-env begin end)
         ;; content.
         (let ((buffer-undo-list t)
               (inhibit-modification-hooks t))
+          (when (and (fboundp 'erm-wait-for-parse)
+                     (derived-mode-p 'enh-ruby-mode))
+            (erm-wait-for-parse))
           ;; Some versions of cc-mode fail when inserting snippet
           ;; content in a narrowed buffer, so make sure to insert
           ;; before narrowing.  Furthermore, call before and after
```

(putting the `erm-wait-for-parse` call earlier works too)

**Comment by zenspider** [comment]
@zw963  / @npostavs can one of you submit a PR to joaotavora/yasnippet with the patch above?

Also makes me wonder if there isn't some sort of generalized protocol to say "this mode needs to finish X before you can do things".

**Comment by npostavs** [comment]
I'm the current yasnippet maintainer, so no specific need for a PR. But I'm not entirely sure putting in this mode-specific hack into yasnippet makes sense.

> Also makes me wonder if there isn't some sort of generalized protocol to say "this mode needs to finish X before you can do things".

For modes using `syntax-propertize-function`, `(syntax-propertize POS)` will wait until the propertize function finishes running up to `POS`. But it tends to be a bad thing if it takes a noticeable amount of time (because it's called by font locking and movement functions). 

Would it make sense to do the waiting in `enh-ruby-mode`'s before-change-function?

**Comment by npostavs** [comment]
> Would it make sense to do the waiting in `enh-ruby-mode`'s before-change-function?

E.g., the following patch fixes it for me, and doesn't break any enh-ruby-mode tests:
```diff
--- i/enh-ruby-mode.el
+++ w/enh-ruby-mode.el
@@ -462,6 +462,10 @@ (define-derived-mode enh-ruby-mode prog-mode "EnhRuby"
   (setq indent-tabs-mode            enh-ruby-indent-tabs-mode)
   (setq imenu-create-index-function 'enh-ruby-imenu-create-index)
 
+  (add-hook 'before-change-functions
+            (lambda (_beg _end) (erm-wait-for-parse))
+            nil t)
+
   (if enh-ruby-add-encoding-comment-on-save
     (add-hook 'before-save-hook 'enh-ruby-mode-set-encoding nil t))
 ```

**Comment by zenspider** [comment]
That DOES break for me... but it's `enh-ruby-toggle-block/do` which has always been a bit sketchy.

Can you tell me how you ran your tests?

**Comment by npostavs** [comment]
Oh, `enh-ruby-toggle-block/do` does fail for me too, I didn't notice because `rake test:elisp` returns a successful exit status in this case and I didn't scroll down.

**Comment by zenspider** [comment]
Yeah. I don't think that's actually on your change... the test has been flaky. See #132. 

I have no idea how to dig into that one.. I should probably mark it as skip or something for now. It's noise.

**Comment by zenspider** [comment]
`enh-ruby-toggle-block/do` has been fixed by @thiagoa so fixing this should be more possible now... however, that `add-hook` block above causes SEVERE slowdowns and had to be removed.

**Comment by zw963** [comment]
Hi,  @zenspider, i meet a new issue which only happen on enh-ruby-mode with yasnippets, could you please check [this]?(https://github.com/joaotavora/yasnippet/issues/969#issuecomment-843397571)

BTW: i am not sure about this is  ``enh-ruby-mode` issue or `yasnippet`, because this issue not happen on 0.13, will happen on 0.14 with enh-ruby-mode only, ruby-mode always work.

Thank you.
